--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -77,6 +77,8 @@
 
 struct net_device * (*get_tnl_dev)(uint32_t tops_crsn) = NULL;
 EXPORT_SYMBOL(get_tnl_dev);
+u32 (*get_tport_etype)(struct sk_buff *skb) = NULL;
+EXPORT_SYMBOL(get_tport_etype);
 
 void mtk_w32(struct mtk_eth *eth, u32 val, unsigned reg)
 {
@@ -1367,6 +1369,8 @@
 	struct mtk_tx_dma_v2 *desc = txd;
 	u64 addr64 = 0;
 	u32 data = 0;
+	u8 tport = 0;
+	u8 etype = 0;
 
 	if (!info->qid && mac->id)
 		info->qid = MTK_QDMA_GMAC2_QID;
@@ -1393,6 +1397,17 @@
 	trace_printk("[%s] skb_shinfo(skb)->nr_frags=%x HNAT_SKB_CB2(skb)->magic=%x txd4=%x<-----\n",
 		     __func__, skb_shinfo(skb)->nr_frags, HNAT_SKB_CB2(skb)->magic, data);
 #endif
+	if (get_tport_etype && skb) {
+		u32 result = get_tport_etype(skb);
+
+		tport = (result & 0xFF00) >> 8;
+		etype = result & 0x00FF;
+	}
+
+	if (tport != 0) {
+		data &= 0xFFFFFFF0;
+		data |= tport;
+	}
 	WRITE_ONCE(desc->txd4, data);
 
 	data = 0;
@@ -1414,7 +1429,10 @@
 	WRITE_ONCE(desc->txd6, data);
 
 	WRITE_ONCE(desc->txd7, 0);
-	WRITE_ONCE(desc->txd8, 0);
+	if (etype != 0)
+		WRITE_ONCE(desc->txd8, etype << 8);
+	else
+		WRITE_ONCE(desc->txd8, 0);
 }
 
 static void mtk_tx_set_dma_desc(struct sk_buff *skb, struct net_device *dev, void *txd,
