--- a/config.c
+++ b/config.c
@@ -266,7 +266,7 @@
 	    errno = 0;
 	    switch(global_switches[i].type) {
 	    case T_INT:
-		* ((int *)global_switches[i].address) = strtoi(s,0,0);
+		* ((int *)global_switches[i].address) = (int)strtol(s,0,0);
 		break;
 	    case T_UINT:
 		* ((unsigned int *)global_switches[i].address) = strtoui(s,0,0);
--- a/floppyd_io.c
+++ b/floppyd_io.c
@@ -470,7 +470,7 @@
 	  p++;
 	*drive = 0;
 	if(*p >= '0' && *p <= '9')
-	  *drive = strtoi(p, &p, 0);
+	  *drive = (int)strtol(p, &p, 0);
 
 	*display = strdup(newname);
 
