--- a/include/image.mk
+++ b/include/image.mk
@@ -441,7 +441,7 @@ else
 endif
 
 FW_ENCRYPT = $(if $(CONFIG_MTK_KERNEL_ENCRYPTION),$(wildcard $(ROE_KEY_DIR)/$(ROE_KEY_NAME).key),)
-ROOTFS_ENCRYPT = $(if $(ROE_KEY_DIR),$(wildcard $(ROE_KEY_DIR)/$(ROE_KEY_NAME).key),)
+ROOTFS_ENCRYPT = $(if $(CONFIG_MTK_ROOTFS_ENCRYPTION),$(wildcard $(ROE_KEY_DIR)/$(ROE_KEY_NAME).key),)
 
 DEVICE_CHECK_FIT_KEY = $(if $(wildcard $(FIT_KEY_DIR)/$(FIT_KEY_NAME).key),install-images,install-disabled)
 DEVICE_CHECK_FIT_DIR = $(if $(FIT_KEY_DIR),$(DEVICE_CHECK_FIT_KEY),install-disabled)
